### PR TITLE
Use Serial as identifier for Application + Tidying up

### PR DESCRIPTION
**Pull request by nmadruga**
While I was working on issue #45 and merging many PRs into it, I found out a few improvements to be added. So break down in this PR with more tidying up changes - and will base this one for the #45 related changes. : )

## Changes
- Use `watch` flag on script `yarn generate` to make it continue checking for changes on graphql queries/mutation/fragments to update the generated ` graphql file.
- Update `cache.ts` file using the `serial` as the identifier key in the Application table (for cache management).
- Add fragments for each entity and use in existing queries/mutations (Attention: This may cause different behaiviour in exisiting functionalities. Please revisit all existing ones and let me know if any problem is happening).
- Update `getApplication` and `updateApplication` query/mutation to use the `serial` to identify the entity.
- Update `ApplicationSummary` to receive from `ApplicationOverview` elements using new hook `getResponsesAndElements` instead of specific one `useLoadSummary` - which is now removed.

## Existing problems:
- When creating the application the elements aren't loaded - I think it is because we haven't got the `skip` on the query added to master yet. Will need to investigate further.
- Responses aren't showing correctly on the application:
Screen Shot 2020-11-12 at 11 01 45 AM

- ~~The progress bar is not showing the pages here. Will need to investigate too.~~ 
  - Problem found - missing `index` in fragment.

### CODE REVIEW

File: src/cache.ts [hunk 1]
```diff
@@ -3,6 +3,9 @@ import { TemplateElementsConnection } from './utils/generated/graphql'
 
 const cache = new InMemoryCache({
     typePolicies: {
+      Application: {
+        keyFields: ['serial']
```
Comment (andreievg) [inline src/cache.ts:7]: who does this need to be done ?

Comment (nmadruga) [reply src/cache.ts:7]: This was done while I was trying to fix what was causing the cache management to not work properly. The problem happening was that after an application was fetched (onLoadApplication), regardless being present in the cache, the App was trying to fetch from the server again for the same application because we are using the `serial` to find it. The cache doesn't have a way to check for other fields in an entity that is already in the cache just by its identifier (aka `keyField`). Since we are using the `serial` to query for applications at all places it makes more sense to use it as the identifier in the cache. And now the Apollo cache is correctly getting the application from cache the 2nd time.

**Comment by andreievg** [comment]
@nmadruga 
> This was done while I was trying to fix what was causing the cache management to not work properly. The problem happening was that after an application was fetched (onLoadApplication), regardless being present in the cache, the App was trying to fetch from the server again for the same application because we are using the serial to find it. The cache doesn't have a way to check for other fields in an entity that is already in the cache just by its identifier (aka keyField). Since we are using the serial to query for applications at all places it makes more sense to use it as the identifier in the cache. And now the Apollo cache is correctly getting the application from cache the 2nd time.

I saw that ID was added to application fragment as one of the later commits, could it have been that application wasn't stored and retrieved from cache because ID was missing ? I would have thought that Apollo Client would allow queries from cache by other fields (rather then just the one that was queried originally)

File: src/components/Application/ApplicationSummary.tsx [hunk 2]
```diff
@@ -79,4 +73,19 @@ const ApplicationSummary: React.FC<ApplicationSummaryProps> = ({
...
+  return question.category === TemplateElementCategory.Question ? (
+    <Form.Field>
+      <Header as="h3" content={question.title} />
+      <Input disabled error={!response}>
+        {response?.text}
+      </Input>
+    </Form.Field>
+  ) : question.elementTypePluginCode !== 'pageBreak' ? (
+    <Form.Field>
+      <Header as="h5" content={question.title} />
```
Comment (andreievg) [inline src/components/Application/ApplicationSummary.tsx:86]: was wondering what this type of element would be ?

File: src/components/Application/ApplicationSummary.tsx [hunk 3]
```diff
@@ -1,57 +1,49 @@
...
-      </Form.Field>
-    ) : null
-  }
+  const sectionsAndElements = Object.entries(elementsState).reduce(
+    (sectionGroups: SectionGroups, [code, item]) => {
+      const group = sectionGroups[item.section] || {}
+      group[code] = item
+      sectionGroups[item.section] = group
+      return sectionGroups
+    },
```
Comment (andreievg) [inline src/components/Application/ApplicationSummary.tsx:35]: I think if we are doing preliminary reshaping like this, it would be great to put all of the required info in the new object (so we do the minimum amount of logic in jsx)

Also it maybe easier to deal with arrays (rather than objects) in jsx.

i.e. if sectionsAndElements was:

```
[
 {
  section: {full section info}
  elements: [ 
      {
          element: {code etc}
          value: {  }
       }
   ]
 }
]
```

It's also easier to sort arrays, by the way are the elements sorted by index within the section ?

File: src/utils/hooks/useGetResponsesAndElementState.tsx [hunk 4]
```diff
@@ -108,8 +102,11 @@ const useGetResponsesAndElementState = (props: { serialNumber: string }) => {
     // const isValid = evaluateExpression(element.validation, evaluationParameters)
     const results = await Promise.all([isEditable, isRequired, isVisible])
     const evaluatedElement = {
-      code: element.code,
       id: element.id,
+      code: element.code,
+      title: element.title,
+      elementTypePluginCode: element.elementTypePluginCode,
+      section: element.section as number,
```
Comment (andreievg) [inline src/utils/hooks/useGetResponsesAndElementState.tsx:109]: its not really part of this PR, but was wondering if we should put the response value here too, and was wondering why we would need 3 types  of return rather then just elementsState, in this hook.

**Review by andreievg — CHANGES_REQUESTED** [review]
Mainly it's looking pretty good, a few comments re application summary page

File: src/containers/Application/ApplicationOverview.tsx [hunk 5]
```diff
@@ -10,23 +11,37 @@ const ApplicationOverview: React.FC = () => {
...
+    serialNumber: serialNumber as string,
+  })
+
+  const {
+    error: responsesError,
+    loading: responsesLoading,
+    responsesByCode,
+    responsesFullByCode,
+    elementsState,
+  } = useGetResponsesAndElementState({
```
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply src/containers/Application/ApplicationOverview.tsx:24]: I already did this in my one: #112

[the remark this answers is not in the document]
Comment (nmadruga) [reply src/containers/Application/ApplicationOverview.tsx:24]: Cool. So should be sorted when we merge those together. I might merge #112 to this branch to sort out conflicts.

File: src/components/Application/ApplicationSummary.tsx (hunk 2 above)
Comment (nmadruga) [reply src/components/Application/ApplicationSummary.tsx:86]: Not sure if I understood your question. Are you asking about the `elementTypePluginCode`?

File: src/containers/Application/ApplicationOverview.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply src/containers/Application/ApplicationOverview.tsx:24]: Yeah. #112 should be ready to merge if you're happy -- have addressed your review comments.

File: src/components/Application/ApplicationSummary.tsx (hunk 3 above)
Comment (nmadruga) [reply src/components/Application/ApplicationSummary.tsx:35]: Doing this now!

File: src/containers/Application/ApplicationOverview.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (nmadruga) [reply src/containers/Application/ApplicationOverview.tsx:26]: Ok! I have merged #112 with this PR. I noticed a problem with using this `isReady` now, but I'll put in another issue so we can merge this issues soon!

File: src/components/Application/ApplicationSummary.tsx (hunk 3 above)
Comment (nmadruga) [reply src/components/Application/ApplicationSummary.tsx]: Have added this suggestion. Please have a look see if you like. I have moved to use arrays as you suggested. And I'm using a sort on the sections by index to keep in the correct order.

File: src/utils/hooks/useGetResponsesAndElementState.tsx (hunk 4 above)
Comment (nmadruga) [reply src/utils/hooks/useGetResponsesAndElementState.tsx:110]: Not sure Andrei, I think we need to add this improvement on another issue really. I'm also not sure if we need all 3 objects, but so far they have been in use, so let's try to tidy up more as we go.

File: src/components/Application/ApplicationSummary.tsx (hunk 2 above)
Comment (andreievg) [reply src/components/Application/ApplicationSummary.tsx:55]: The template element that's not a question or page break is just displayed as h5 header, was wondering if it needs to be there

File: src/utils/hooks/useGetResponsesAndElementState.tsx (hunk 4 above)
Comment (andreievg) [reply src/utils/hooks/useGetResponsesAndElementState.tsx:110]: cools

File: src/components/Application/ApplicationSummary.tsx (hunk 3 above)
Comment (andreievg) [reply src/components/Application/ApplicationSummary.tsx]: looks great thanks !

File: src/containers/Application/ApplicationOverview.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (andreievg) [reply src/containers/Application/ApplicationOverview.tsx:26]: cool, I realised that we can use loading response from LoadApplication for this (but still there seems to be an improvement to be done in useLoadApplications to only load it once)

**Review by andreievg — APPROVED** [review]
Looks good, thanks Nicole.

Still there is a comment regarding keyFields 'serial' that i've made, be good if you can check it out sometime
 https://github.com/openmsupply/application-manager-web-app/pull/110#issuecomment-727645790

File: src/components/Application/ApplicationSummary.tsx (hunk 2 above)
Comment (nmadruga) [reply src/components/Application/ApplicationSummary.tsx:55]: Oh, that was just to replace what will be the `SumaryView` of other elements types in the future. While its not implemented in the plugins its showing like that.
